src/helpers.cpp

```diff
@@ -147,8 +146,212 @@ Vec3 Triangulate2(const SE3 &g12, const Vec2 &xc1, const Vec2 &xc2) {
+
+  Eigen::JacobiSVD<Eigen::Matrix<double, 2, 3>> svd(B, Eigen::ComputeFullV);
+  Mat3 V = svd.matrixV();
+  Vec3 n_prime_hat = V.col(1);
+
+  Vec3 m0_prime = m0 - m0.dot(n_prime_hat) * n_prime_hat;
+  Vec3 m1_prime = m1 - m1.dot(n_prime_hat) * n_prime_hat;
+
+  Vec3 Rf0_prime = m0_prime;
+  Vec3 f1_prime = m1_prime;
+
+  Vec3 z = f1_prime.cross(Rf0_prime);
+
+  x = ((z.dot(t21.cross(Rf0_prime))) / pow(z.norm(),2)) * f1_prime;
+
+  if(!check_cheirality(z, t21, f1_prime, Rf0_prime) ||
+    !check_angular_reprojection(m0, Rf0_prime, m1, f1_prime) ||
+    !check_parallax(Rf0_prime, f1_prime))
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool LinfAngular(const SE3 &g12, const Vec2 &xc0, const Vec2 &xc1, Vec3 &x) {
+
+  // Initalize the Rotation and Translation Matricies
+  Vec3 t12{g12.T()};
+  Mat3 R12{g12.R()};
+  Mat3 R21{R12.transpose()};
+  Vec3 t21{-1 * R12.transpose() * t12};
+
+  // Create homogeneous coordinates
+  Vec3 f0{xc0(0), xc0(1), 1.0};
+  f0.normalize();
+  Vec3 f1{xc1(0), xc1(1), 1.0};
+  f1.normalize();
+
+  Vec3 m0{R21 * f0};
+  Vec3 m1{f1};
+
+  Vec3 m0_hat = m0 / m0.norm();
+  Vec3 m1_hat = m1 / m1.norm();
+
+  Vec3 n_a = (m0_hat + m1_hat).cross(t21);
+  Vec3 n_b = (m0_hat - m1_hat).cross(t21);
+
+  Vec3 n_prime_hat = n_a.norm() >= n_b.norm() ? n_a : n_b;
+
+  Vec3 m0_prime = m0 - m0.dot(n_prime_hat) * n_prime_hat;
+  Vec3 m1_prime = m1 - m1.dot(n_prime_hat) * n_prime_hat;
+
+  Vec3 Rf0_prime = m0_prime;
+  Vec3 f1_prime = m1_prime;
+
+  Vec3 z = f1_prime.cross(Rf0_prime);
+
+  x = ((z.dot(t21.cross(Rf0_prime))) / pow(z.norm(),2)) * f1_prime;
+
+  if(!check_cheirality(z, t21, f1_prime, Rf0_prime) ||
+    !check_angular_reprojection(m0, Rf0_prime, m1, f1_prime) ||
+    !check_parallax(Rf0_prime, f1_prime))
+  {
+    return false;
+  }
+
+  return true;
+}
+
+
+bool check_cheirality(const Vec3 &z, const Vec3 &t, const Vec3 &f1_prime, const Vec3 &Rf0_prime)
+{
+
+  float lambda0 = z.dot(t.cross(f1_prime)) / pow(z.norm(), 2);
+  float lambda1 = z.dot(t.cross(Rf0_prime)) / pow(z.norm(), 2);
+
+  if(lambda0 <= 0 || lambda1 <= 0)
+  {
+    LOG(ERROR) << "[ERROR] cheirality error in triangulation. lambda0=" << lambda0 << ", lamba1=" << lambda1;
+    return false;
+  }
+
+  return true;
+}
+
+
+bool check_angular_reprojection(const Vec3 &Rf0, const Vec3 &Rf0_prime, const Vec3 &f1, const Vec3 &f1_prime)
+{
+
+  float theta0 = acos(Rf0.dot(Rf0_prime) / (Rf0.norm() * Rf0_prime.norm()));
+  float theta1 = acos(f1.dot(f1_prime) / (f1.norm() * f1_prime.norm()));
+
+  float max_theta = std::max(theta0, theta1);
+
+  if(max_theta > 0.01)
+  {
+    LOG(ERROR) << "[ERROR] angular reprojection error in triangulation";
+    return false;
+  }
+  return true;
+}
+
+bool check_parallax(const Vec3 &Rf0_prime, const Vec3 &f1_prime)
+{
+
+  float beta = acos(f1_prime.dot(Rf0_prime) / (f1_prime.norm() * Rf0_prime.norm()));
+
+  if(beta < 0.1)
+  {
+    LOG(ERROR) << "[ERROR] parallax error in triangulation " << beta;
```

Comment:
I think chierality, parallax, and angular reprojection errors should be `LOG(WARNING)` instead of `LOG(ERROR)`. XIVO just falls back to estimating depth using only the subfilter and goes on. In many perfectly valid uses cases, we should be throwing those errors.

For the record, though, the definition of error vs. warning is up to us. But this is the [web page with the definitions that I was looking at](https://www.esrf.fr/computing/scientific/FIT2D/FIT2D_REF/node18.html).